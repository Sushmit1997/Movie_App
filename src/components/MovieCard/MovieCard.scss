.movieCardMainContainer{
    display: flex;
    flex-direction: column;
    margin-bottom: 15px;
    padding: 10px;
    background-color: #ffffff;
    border-radius: 4px;
    color: #2c2c54;
    transition: box-shadow  0.2s ease-in;
    width: 500px;

    &:hover{
      box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19);
    }
}

.movieCardInnerDiv{
    display: flex;
}


.movieCardImage{
    width: 150px;
    height: 250px;

}

.movieCardInfoSectionContainer{
    width: 100%;
    display: flex;
    flex-direction: column;
    position: relative;
}

.movieCardInfoSection{
    margin-left: 35px;
    margin-top: 25px;
    display: flex;
    flex-direction: column;
}

.movieCardInfoSection span:first-child{
    font-size: 30px;
}

.movieCardInfoSection span {
    margin-bottom: 15px;
    word-break: break-all;
    text-align: left;
}

.heartIconDiv{
    position: absolute;
    display: flex;
    justify-content: flex-end;
    right: 20px;
}

.fas.fa-heart{
    transition:color 0.1s ease-in;
    font-size: 18px;
    transition: font-size  0.15s ease-in;
    
    &:hover{
        color:red;
        font-size:20px

    }
}

.fas.fa-heart.red{
    color: red;
}

.movieDirectorSpan{
    cursor: pointer;
    transition: font-weight 0.05s ease-in;

    &:hover{
        font-weight: 600;
    }
}


//Accordion styling

.movieAccordionContainer{
}

.accordion-title {
    display: flex;
    flex-direction: row;
    justify-content: flex-end;
    margin-top: auto;
  }
  
  
  .accordion-title,
  .accordion-content {
    padding: 1rem;
  }

  .movieDetailLoadingDiv{
      display: flex;
      justify-content: center;
  }
  
  .accordion-content {
      display: flex;
      flex-direction: column;


    .accordion-ul{
        padding-left: 0px;
        list-style: none;

        li{
            text-align: left;
            margin-bottom: 10px;

            span{
                font-weight: bold;
            }
        }
    }
  }


