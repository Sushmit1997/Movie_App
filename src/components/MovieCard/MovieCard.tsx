import React, { useState, useEffect } from 'react'
import { IMovieData } from '../../interfaces/IMovieData'
import { useSnackbar } from 'notistack'
import { makeStyles } from '@material-ui/core/styles'
import Modal from '@material-ui/core/Modal'
import Backdrop from '@material-ui/core/Backdrop'
import Fade from '@material-ui/core/Fade'
import CircularProgress from '@material-ui/core/CircularProgress'
import { get } from '../../utils/http'

interface Props {
	movieInfo: IMovieData
	key: number
}

const useStyles = makeStyles((theme) => ({
	modal: {
		display: 'flex',
		alignItems: 'center',
		justifyContent: 'center',
	},
	paper: {
		backgroundColor: theme.palette.background.paper,
		padding: theme.spacing(2, 4, 3),
		width: '60%',
	},
}))

const MovieCard: React.FC<Props> = (props: Props) => {
	const { movieInfo } = props

	const { enqueueSnackbar, closeSnackbar } = useSnackbar()

	const [movieFav, setMovieFav] = useState(false)
	const [isActive, setIsActive] = useState(false)
	const [movieDetail, setMovieDetail]: any = useState('')
	const [movieDetailLoading, setMovieDetailLoading] = useState(true)
	const [modal, setModal] = useState(false)

	const toggle = (e) => {
		e.stopPropagation()
		setModal(!modal)
	}

	const handleFakePromiseFav = () => {
		var randomBoolean = Math.random() < 0.5
		return new Promise((resolve, reject) => {
			{
				randomBoolean ? resolve(randomBoolean) : reject(randomBoolean)
			}
		})
	}

	const handleFavMovie = (e) => {
		e.stopPropagation()
		handleFakePromiseFav()
			.then((response) => {
				setTimeout(() => {
					setMovieFav(!movieFav)
					movieFav
						? enqueueSnackbar(`${movieInfo.title} removed from favourites.`, {
								variant: 'warning',
								anchorOrigin: {
									vertical: 'bottom',
									horizontal: 'right',
								},
						  })
						: enqueueSnackbar(`${movieInfo.title} added to favourites.`, {
								variant: 'success',
								anchorOrigin: {
									vertical: 'bottom',
									horizontal: 'right',
								},
						  })
				}, 500)
			})
			.catch((error) => {
				setTimeout(() => {
					enqueueSnackbar(`Something went wrong! Please try again.`, {
						variant: 'error',
						anchorOrigin: {
							vertical: 'bottom',
							horizontal: 'right',
						},
					})
				}, 500)
			})
	}

	const fetchMovieDetails = () => {
		setIsActive(!isActive)
		// {
		// 	!isActive &&
		// 		movieDetail === '' &&
		// 		get(
		// 			`https://imdb-api.com/en/API/Title/k_patmt9tu/${movieInfo.id}/FullActor,Ratings,`
		// 		).then((response: any) => {
		// 			setMovieDetail(response.data)
		// 			setMovieDetailLoading(false)
		// 		})
		// }
	}

	const handleDirectorClick = (e) => {
		e.stopPropagation()
		toggle(e)
	}

	const classes = useStyles()

	return (
		<div
			data-testid={`movieCard`}
			className='movieCardMainContainer'
			onClick={fetchMovieDetails}
		>
			<div className='movieCardInnerDiv'>
				<div className='movieCardImageDiv'>
					<img className='movieCardImage' src={movieInfo.image}></img>
				</div>
				<div className='movieCardInfoSectionContainer'>
					<div className='heartIconDiv' onClick={(e) => handleFavMovie(e)}>
						<i className={`fas fa-heart ${movieFav ? 'red' : ''} `}></i>
					</div>
					<div className='movieCardInfoSection'>
						<span>
							{movieInfo.title} {movieInfo.description}{' '}
						</span>
						<span> {movieInfo.type}</span>
						<span
							className='movieDirectorSpan'
							onClick={(e) => handleDirectorClick(e)}
							style={{ width: 'fit-content' }}
						>
							{movieInfo.director}
						</span>
					</div>
					<div
						className='accordion-title'
						onClick={() => setIsActive(!isActive)}
					>
						<div>{isActive ? '-' : '+'}</div>
					</div>
				</div>
			</div>
			{isActive && (
				<div className='movieAccordionContainer'>
					<div className='accordion'>
						<div className='accordion-item'>
							{isActive && (
								<div className='accordion-content'>
									{movieDetailLoading ? (
										<div className='movieDetailLoadingDiv'>
											<CircularProgress color='primary' />
										</div>
									) : (
										<ul className='accordion-ul'>
											{movieDetail.imDbRating === null && (
												<li>
													{' '}
													If you see no data, probably api limit is reached.
												</li>
											)}

											<li>
												{' '}
												<span>Awards:</span> {movieDetail.awards}
											</li>
											<li>
												{' '}
												<span>Budget:</span>{' '}
												{movieDetail.boxOffice !== null &&
													movieDetail.boxOffice}
											</li>
											<li>
												<span>Imdb Rating:</span> {movieDetail.imDbRating}{' '}
												<span>Voted by:</span> {movieDetail.imDbRatingVotes}
											</li>
											<li>
												<span>Plot:</span> {movieDetail.plot}
											</li>
											<li>
												<span>Released Date:</span> {movieDetail.releaseDate}
											</li>
										</ul>
									)}
								</div>
							)}
						</div>
					</div>
				</div>
			)}

			<Modal
				aria-labelledby='transition-modal-title'
				aria-describedby='transition-modal-description'
				className={classes.modal}
				open={modal}
				onClose={(e) => toggle(e)}
				closeAfterTransition
				BackdropComponent={Backdrop}
				BackdropProps={{
					timeout: 500,
				}}
			>
				<Fade in={modal}>
					<div className={classes.paper}>
						<h2 id='transition-modal-title'>{movieInfo.directorInfo.name}</h2>
						<div className='directorInfoInnerContainer'>
							<img
								className='directorImage'
								src={movieInfo.directorInfo.image}
								alt={`Picture of ${movieInfo.directorInfo.name}`}
							></img>
							<p id='transition-modal-description'>
								{movieInfo.directorInfo.description}
							</p>
						</div>
					</div>
				</Fade>
			</Modal>
		</div>
	)
}

export default MovieCard
