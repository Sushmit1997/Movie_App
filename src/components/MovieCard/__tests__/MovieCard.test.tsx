import MovieCard from '../MovieCard'
import { render, screen, cleanup } from '@testing-library/react'
import movieData from '../../../utils/dummyData'
import '@testing-library/jest-dom/extend-expect'
import React from 'react'
import { SnackbarProvider } from 'notistack'
import renderer from 'react-test-renderer'

test('Should render Inception text', () => {
	render(
		<SnackbarProvider maxSnack={3}>
			{' '}
			<MovieCard key={1} movieInfo={movieData[0]} />{' '}
		</SnackbarProvider>
	)
	const movieCardElement = screen.getByTestId('movieCard')
	expect(movieCardElement).toHaveTextContent('Inception')
})

test('Should render Avengers Text', () => {
	render(
		<SnackbarProvider maxSnack={3}>
			{' '}
			<MovieCard key={1} movieInfo={movieData[3]} />{' '}
		</SnackbarProvider>
	)
	const movieCardElement = screen.getByTestId('movieCard')
	expect(movieCardElement).toHaveTextContent('Avengers')
})

test('<MovieCard/> renderes with correct text', () => {
	const tree = renderer
		.create(
			<SnackbarProvider maxSnack={3}>
				{' '}
				<MovieCard key={1} movieInfo={movieData[2]} />{' '}
			</SnackbarProvider>
		)
		.toJSON()
	expect(tree).toMatchSnapshot()
})
