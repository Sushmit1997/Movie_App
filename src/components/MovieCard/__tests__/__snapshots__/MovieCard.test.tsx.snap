// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard/> renderes with correct text 1`] = `
Array [
  " ",
  <div
    className="movieCardMainContainer"
    data-testid="movieCard"
    onClick={[Function]}
  >
    <div
      className="movieCardInnerDiv"
    >
      <div
        className="movieCardImageDiv"
      >
        <img
          className="movieCardImage"
          src="https://imdb-api.com/images/original/MV5BMTc5MDE2ODcwNV5BMl5BanBnXkFtZTgwMzI2NzQ2NzM@._V1_Ratio0.7273_AL_.jpg"
        />
      </div>
      <div
        className="movieCardInfoSectionContainer"
      >
        <div
          className="heartIconDiv"
          onClick={[Function]}
        >
          <i
            className="fas fa-heart  "
          />
        </div>
        <div
          className="movieCardInfoSection"
        >
          <span>
            Avengers Endgame
             
            (2019)
             
          </span>
          <span>
             
            Superhero /Action
          </span>
          <span
            className="movieDirectorSpan"
            onClick={[Function]}
            style={
              Object {
                "width": "fit-content",
              }
            }
          >
            Joe Russo, Anthony Russo
          </span>
        </div>
        <div
          className="accordion-title"
          onClick={[Function]}
        >
          <div>
            +
          </div>
        </div>
      </div>
    </div>
  </div>,
  " ",
]
`;
